src/node/lattice.rs

```diff
@@ -188,90 +191,81 @@ impl ExecutionLattice {
-            let mut dfs = DfsPostOrder::empty(&*forest);
-            let mut dependencies: HashSet<NodeIndex<u32>> = HashSet::new();
-            let mut add_edges_to: Vec<NodeIndex<u32>> = Vec::new();
-            let mut add_edges_from: Vec<NodeIndex<u32>> = Vec::new();
-            let mut remove_roots: Vec<RunnableEvent> = Vec::new();
-
-            for root in roots.iter() {
-                // Retrieve the index into the lattice.
-                let root_index: NodeIndex<u32> = root.node_index;
-
-                // This function maintains the stack when iterating over the graph so we do not have to
-                // check the same node in the graph twice.
-                dfs.move_to(root_index);
-
-                // Retrieve the next node in DFS order from the forest.
-                while let Some(nx) = dfs.next(&*forest) {
-                    // Check if any of the children of the current node in DFS traversal already have a
-                    // dependency on the node we are trying to add.
-                    let mut dependent_child: bool = false;
-                    for child in forest.neighbors(nx) {
-                        if dependencies.contains(&child) {
-                            dependencies.insert(nx);
-                            dependent_child = true;
-                            break;
+            // Visits each node in the graph once.
+            let mut dfs = DfsPostOrder::empty(Reversed(&*forest));
+            // Caches preceding events to avoid adding redundant dependencies.
+            // For example, A -> C is redundant if A -> B -> C.
+            let mut preceding_events: HashSet<NodeIndex<u32>> = HashSet::new();
+            // The added event depends on these nodes.
+            let mut outgoing_edges: Vec<NodeIndex<u32>> = Vec::new();
+            // Other nodes depend on the added event.
+            let mut incoming_edges: Vec<NodeIndex<u32>> = Vec::new();
+            // These nodes are no longer roots after the added event is inserted into the graph.
+            let mut demoted_roots: Vec<RunnableEvent> = Vec::new();
+            // Iterate through the forest with a DFS from each root to figure out where to add dependency edges.
+            'dfs_roots: for root in roots.iter() {
+                // Begin a DFS at the specified root.
+                dfs.move_to(root.node_index);
+                while let Some(nx) = dfs.next(Reversed(&*forest)) {
+                    // If any of the current node's children precede the added event, then the current node also precedes
+                    // the added event. By induction, the root would also precede the added event, so we can move on to the
+                    // next root.
+                    for child in forest.neighbors_directed(nx, Direction::Incoming) {
+                        if preceding_events.contains(&child) {
+                            preceding_events.insert(nx);
+                            continue 'dfs_roots;
```

Comment:
Found [this Rust syntax](https://doc.rust-lang.org/reference/expressions/loop-expr.html#continue-expressions) surprisingly elegant.

Reply:
Logically, it's the same as before, but instead of keeping a boolean to continue the outer loop, the Rust syntax lets us continue the outer loop directly :)